src/Type/Php/ArrayFilterFunctionReturnTypeExtension.php

```diff
@@ -62,290 +27,7 @@ public function getTypeFromFunctionCall(FunctionReflection $functionReflection,
-				$builder = ConstantArrayTypeBuilder::createEmpty();
-				$optionalKeys = $constantArray->getOptionalKeys();
-				foreach ($constantArray->getKeyTypes() as $i => $keyType) {
-					$itemType = $constantArray->getValueTypes()[$i];
-					[$newKeyType, $newItemType, $optional] = $this->processKeyAndItemType($scope, $keyType, $itemType, $itemVar, $keyVar, $expr);
-					$optional = $optional || in_array($i, $optionalKeys, true);
-					if ($newKeyType instanceof NeverType || $newItemType instanceof NeverType) {
-						continue;
-					}
-					if ($itemType->equals($newItemType) && $keyType->equals($newKeyType)) {
-						$builder->setOffsetValueType($keyType, $itemType, $optional);
-						continue;
-					}
-
-					$builder->setOffsetValueType($newKeyType, $newItemType, true);
-				}
-
-				$results[] = $builder->getArray();
-			}
-
-			return TypeCombinator::union(...$results);
-		}
-
-		[$newKeyType, $newItemType] = $this->processKeyAndItemType($scope, $arrayType->getIterableKeyType(), $arrayType->getIterableValueType(), $itemVar, $keyVar, $expr);
-
-		if ($newItemType instanceof NeverType || $newKeyType instanceof NeverType) {
-			return new ConstantArrayType([], []);
-		}
-
-		return new ArrayType($newKeyType, $newItemType);
-	}
-
-	/**
-	 * @return array{Type, Type, bool}
-	 */
-	private function processKeyAndItemType(MutatingScope $scope, Type $keyType, Type $itemType, Error|Variable|null $itemVar, Error|Variable|null $keyVar, Expr $expr): array
-	{
-		$itemVarName = null;
-		if ($itemVar !== null) {
-			if (!$itemVar instanceof Variable || !is_string($itemVar->name)) {
-				throw new ShouldNotHappenException();
-			}
-			$itemVarName = $itemVar->name;
-			$scope = $scope->assignVariable($itemVarName, $itemType, new MixedType());
-		}
-
-		$keyVarName = null;
-		if ($keyVar !== null) {
-			if (!$keyVar instanceof Variable || !is_string($keyVar->name)) {
-				throw new ShouldNotHappenException();
-			}
-			$keyVarName = $keyVar->name;
-			$scope = $scope->assignVariable($keyVarName, $keyType, new MixedType());
-		}
-
-		$booleanResult = $scope->getType($expr)->toBoolean();
-		if ($booleanResult->isFalse()->yes()) {
-			return [new NeverType(), new NeverType(), false];
-		}
-
-		$scope = $scope->filterByTruthyValue($expr);
-
-		return [
-			$keyVarName !== null ? $scope->getVariableType($keyVarName) : $keyType,
-			$itemVarName !== null ? $scope->getVariableType($itemVarName) : $itemType,
-			!$booleanResult instanceof ConstantBooleanType,
-		];
-	}
-
-	private static function createFunctionName(string $funcName): ?Name
-	{
-		if ($funcName === '') {
-			return null;
-		}
-
-		if ($funcName[0] === '\\') {
-			$funcName = substr($funcName, 1);
-
-			if ($funcName === '') {
-				return null;
-			}
-
-			return new Name\FullyQualified($funcName);
-		}
-
-		return new Name($funcName);
-	}
-
-	/**
-	 * @param self::USE_* $mode
-	 * @return array{list<Arg>, ?Variable, ?Variable}
-	 */
-	private function createDummyArgs(int $mode): array
```

Comment:
Do this refactoring in a separate PR first please.

Reply:
Ok, create a pull request.

Reply:
Cherry-picked https://github.com/phpstan/phpstan-src/pull/3606/